Explain various fields in the benchmark identifier (#6947)
google/iree: benchmarks/README.md

pipeline also runs on pull requests with the `buildkite:benchmark` label,
posting results compared against their base commit as comments.

## Benchmarking philosophy

> Benchmarking and interpret results properly is a delicate task. Typically we can
> pull numbers from various parts of a system, but depending on what we are trying
> to evalulate, those numbers may or may not be relevant. For example, for
> somebody working solely on better kernel code generation, the end-to-end model
> reference latency is unlikely meaningful given it involves lots of runtime
> overhead. Also the environment could vary per benchmark run in uncontrollable
> manner, causing unstability in the results. This is especially true for Android,
> where we face tight compromise between performance and thermal/battery limits.
> Too many aspects can affect the benchmarking results. So before going into
> details, it's worth nothing the general guideline to IREE benchmarking as
> context.

Reviewer: grammar/spelling, reworded:

```suggestion
Benchmarking and interpreting results properly is a delicate task. We can record
metrics from various parts of a system, but depending on what we are trying to
evaluate, those numbers may or may not be relevant. For example, for somebody
working solely on better kernel code generation, the end-to-end model reference
latency is unlikely meaningful given it also includes runtime overhead. The
environment could also vary per benchmark run in uncontrollable ways, causing
instability in the results. This is especially true for Android, where a tight
compromise between performance and thermal/battery limits is made.
Too many aspects can affect the benchmarking results.
```

> The overaching goal for benchmarking here is to track IREE's performance
> progress and guard against regression. So the benchmarks are meant to understand
> the performance of IREE _itself_, not the absolute capability of the exercised
> hardware. In order to fulfill the above goal, we have the following guidelines
> for benchmarking:

Reviewer: typo

```suggestion
The overarching goal for benchmarking here is to track IREE's performance
progress and guard against regression. So the benchmarks are meant to understand
the performance of IREE _itself_, not the absolute capability of the exercised
hardware. In order to fulfill the above goal, we have the following guidelines
for benchmarking:
```

* We choose representative real-world models with varying characteristics.

> * We cover different IREE backends and different modes for each backend so that
>   folks working on different component can find the metrics they need.

Reviewer: grammar

```suggestion
* We cover different IREE backends and different modes for each backend so that
  folks working on different components can find the metrics they need.
```

## Benchmark specification

Each benchmark in IREE has a unique identifier with the following format:

```
<model-name> `[` <model-tag>.. `]` `(` <model-source> `)` <benchmark-mode>..
`with` <iree-driver>
`@` <device-name> `(` <target-architecture> `)`

> ```

Reviewer: This is specific to model inference latency benchmarks. What about https://github.com/google/iree/issues/6161?

Author: Renamed this to "Model benchmark specification" for now. Once we have that we can then have another section and have a proper layout for them.

      * models converted to TensorFlow Lite and imported with `iree-import-tflite`
```

### Model name

> THis field specifies the input model:

Reviewer: capitalization

```suggestion
This field specifies the input model:
```

* `MobileSSD` [[source](https://www.tensorflow.org/lite/performance/gpu#demo_app_tutorials)]:
  Vision model for object detection.
  Characteristics: convolution, feedforward NN.
* `PoseNet` [[source](https://tfhub.dev/tensorflow/lite-model/posenet/mobilenet/float/075/1/default/1)]:
  Vision model for pose estimation.
  Characteristics: convolution, feedforward NN.

### Model tag

This field specifies the model variant. It depends on the model, but here are
some examples:

* `f32`: the model is working on float types.
* `imagenet`: the model takes ImageNet-sized inputs (224x224x3).

### IREE driver

This field specifies the IREE HAL driver:

* [`Dylib`](https://google.github.io/iree/deployment-configurations/cpu-dylib/):
  For CPU via dynamic library. Kernels contain CPU native instructions AOT
  compiled using LLVM. This driver issues workload to the CPU in async
  manner and supports multithreading.
* [`Dylib-Sync`](https://google.github.io/iree/deployment-configurations/cpu-dylib/):
  For CPU via dynamic library. Kernels contain contain CPU native instructions
  AOT compiled using LLVM. This driver issues workload to the CPU in sync

>   manner.

Reviewer: Argh, another place that will need updating when we finally fix the HAL driver setup (https://github.com/google/iree/issues/4298 and others)

For here, we should also be measuring the other CPU configurations that don't use dynamic libraries.

* [`VMVX`](https://github.com/google/iree/issues/5123):
  For CPU via dynamic library. Kernels contain vector-level intrinsics that
  are backed by fast implementations ([WIP](https://github.com/google/iree/issues/5819)).
  This driver issues workload to the CPU in async manner and supports
  multithreading.
* [`Vulkan`](https://google.github.io/iree/deployment-configurations/gpu-vulkan/):
  For GPU via Vulkan. Kernels contain SPIR-V. This driver issues workload to
  the GPU via the Vulkan API.

### Device name and target architecture

> This two fields are tightly coupled. They specify the device and hardware

Reviewer: ```suggestion
These two fields are tightly coupled. They specify the device and hardware
```

target for executing the benchmark.

Right now there are two Android devices:

* `Pixel-4`: Google Pixel 4 running Android 11. The SoC is
  [Snapdragon 855](https://www.qualcomm.com/products/snapdragon-855-plus-and-860-mobile-platform),
  with 1+3+4 ARMv8.2 CPU cores and Adreno 640 GPU.
* `SM-G980F`: Samsung Galaxy S20 running Android 11. The SoC is
  [Exynos 990](https://www.samsung.com/semiconductor/minisite/exynos/products/mobileprocessor/exynos-990/),

>   with 2+2+4 ARMv8.2 CPU cores and Mali G77 MP11 GPU.

Reviewer: Hopefully this gets out of date very quickly...

Author: What do you mean here? These phones will likely stay for the foreseeable future unless there is an issue with them.

Reviewer: "Right now there are two Android devices" - we should be testing on more devices. Both more of the same devices for benchmark queue times and different devices for more variety (x86, iOS, other phone models, etc.).

Maintaining a list of the specific devices we test on with this level of detail should be difficult.

Documentation that will (or should) easily get out of date is a sign that the documentation should be written or generated differently.

Author: +1 to benchmarking on more devices and variants. Hopefully we can have that soon. :)

But even if we have more devices, it's unlikely to scale like other dimensions (models, modes, etc.). I'd imagine having ~10 is already many. For more, need a dedicated person/team probably. 

In the meanwhile, this is still valuable information to provide. Without these SoC characteristics, it's hard to evaluate those benchmarks properly. (Users could look up by themselves too but I feel that argument can extend to many other things documented too.)

Anyway, we can adjust once we have more devices supported.

Therefore the target architectures are:

* `CPU-CPU-ARMv8.2-A`: can benchmark all CPU-based IREE backends and drivers.
* `GPU-Adreno-640`: can benchmark IREE Vulkan with Adreno target triples.
* `GPU-Mali-G77`: can benchmark IREE Vulkan with Mali target triples.

### Benchmark mode

This field is to further specify the benchmark variant, given the same input
model and target architecture. It controls important aspects like:

* `*-core`: specifies the core flavor for CPU.
* `*-thread`: specifies the number of threads for CPU.
* `full-inference`: measures the latency for one full inference. Note that this
  does not include the IREE system initialization time.
* `kernel-execution`: measures only kernel execution latency for GPU. Note that
  this is only possible for feedforward NN models that can be put into one
  command buffer.

`*-core` and `*-thread` together determines the `taskset` mask used for
benchmarking IREE backends and drivers on CPU. For example,

* `1-thread,big-core` would mean `taskset 80`.
* `1-thread,little-core` would mean `taskset 08`.
* `3-thread,big-core` would mean `taskset f0`.
* `3-thread,little-core` would mean `taskset 0f`.

## IREE developer guides

> The following sections are meant for developers working on IREE itself.

Reviewer: That's what I intended for the README in general :P benchmark definitions come with documentation for how to edit those definitions and a brief overview of how they are used. Philosophy and interpretation docs I imagined going elsewhere.

Author: Yeah. I'm back and forth on where to put this. Not sure we want to put it on https://google.github.io/iree/ given this is quite internal development facing (although other interested folks will also check them so worth having documentation like this). In `docs/developers` we already have a few benchmarking related docs for  benchmarking flow, which is even lower level than this. So I figured right now it's better to get the content out instead of finding the perfect file/dir organization for it. So put it here. ;-P 

Splitted into a separate `dashboard.md` file now in f984896.

Author: f984896 is purely splitting the file. Comments were addressed in 53c0103.
